lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -550,200 +427,244 @@ export class UserAgentApplication {
+    }
   }
 
   /**
    * @hidden
-   * Helper function to loginPopup
+   * @ignore
+   * Helper function to acquireToken
    *
-   * @param account
-   * @param request
-   * @param resolve
-   * @param reject
-   * @param scopes
    */
-  private loginPopupHelper(account: Account, resolve: any, reject: any, request?: AuthenticationParameters, scopes?: Array<string>) {
-    if (!scopes) {
-      scopes = [this.clientId];
-    }
-    const scope = scopes.join(" ").toLowerCase();
-
-    // Generate a popup window
-    const popUpWindow = this.openWindow("about:blank", "_blank", 1, this, resolve, reject);
-    if (!popUpWindow) {
-      // We pass reject in openWindow, we reject there during an error
-      return;
+  private acquireTokenHelper(account: Account, interactionType: InteractionType, isLoginCall: boolean, request?: AuthenticationParameters, scopes?: Array<string>, resolve?: any, reject?: any): void {
+    // Track the acquireToken progress
+    if (isLoginCall) {
+      this.loginInProgress = true;
+    } else {
+      this.acquireTokenInProgress = true;
     }
 
-    // Track login progress
-    this.loginInProgress = true;
-
-    // Resolve endpoint
-    this.authorityInstance.resolveEndpointsAsync().then(() => {
-      let serverAuthenticationRequest = new ServerRequestParameters(this.authorityInstance, this.clientId, scopes, ResponseTypes.id_token, this.getRedirectUri(), request && request.state);
-
-      // populate QueryParameters (sid/login_hint/domain_hint) and any other extraQueryParameters set by the developer;
-      serverAuthenticationRequest = this.populateQueryParams(account, request, serverAuthenticationRequest);
+    const scope = scopes ? scopes.join(" ").toLowerCase() : this.clientId.toLowerCase();
 
-      this.updateCacheEntries(serverAuthenticationRequest, account, window.location.href);
+    let serverAuthenticationRequest: ServerRequestParameters;
+    const acquireTokenAuthority = (!isLoginCall && request.authority) ? AuthorityFactory.CreateInstance(request.authority, this.config.auth.validateAuthority) : this.authorityInstance;
 
-      // Cache the state, nonce, and login request data
-      this.cacheStorage.setItem(Constants.loginRequest, window.location.href, this.inCookie);
-      this.cacheStorage.setItem(Constants.loginError, "");
+    let popUpWindow: Window;
+    if (interactionType === Constants.interactionTypePopup) {
+      // Generate a popup window
+      popUpWindow = this.openWindow("about:blank", "_blank", 1, this, resolve, reject);
+      if (!popUpWindow) {
+        // We pass reject in openWindow, we reject there during an error
+        return;
+      }
+    }
 
-      this.cacheStorage.setItem(Constants.nonceIdToken, serverAuthenticationRequest.nonce, this.inCookie);
+    acquireTokenAuthority.resolveEndpointsAsync().then(() => {
+      // On Fulfillment
+      let responseType: string;
```

Comment:
const responseType = isLoginCall ?  responseType = ResponseTypes.id_token : responseType = this.getTokenType(account, scopes, false);